cmd/tsdb/main.go

```diff
@@ -188,57 +329,110 @@ func (b *writeBenchmark) ingestScrapes(lbls []labels.Labels, scrapeCount int) (u
 	return total, nil
 }
 
-func (b *writeBenchmark) ingestScrapesShard(metrics []labels.Labels, scrapeCount int, baset int64) (uint64, error) {
+func (b *writeBenchmark) ingestScrapesShard(metrics []labels.Labels, m3dbMetrics []ident.TagsIterator, scrapeCount int, baset int64) (uint64, error) {
 	ts := baset
-
-	type sample struct {
-		labels labels.Labels
-		value  int64
-		ref    *uint64
-	}
-
-	scrape := make([]*sample, 0, len(metrics))
-
-	for _, m := range metrics {
-		scrape = append(scrape, &sample{
-			labels: m,
-			value:  123456789,
-		})
-	}
 	total := uint64(0)
 
-	for i := 0; i < scrapeCount; i++ {
-		app := b.storage.Appender()
-		ts += timeDelta
+	if b.useM3DB() {
+		type sample struct {
+			tags  ident.TagsIterator
+			value int64
+			ref   *uint64
+		}
 
-		for _, s := range scrape {
-			s.value += 1000
+		scrape := make([]*sample, 0, len(metrics))
 
-			if s.ref == nil {
-				ref, err := app.Add(s.labels, ts, float64(s.value))
-				if err != nil {
-					panic(err)
-				}
-				s.ref = &ref
-			} else if err := app.AddFast(*s.ref, ts, float64(s.value)); err != nil {
+		for _, m := range m3dbMetrics {
+			scrape = append(scrape, &sample{
+				tags:  m,
+				value: 123456789,
+			})
+		}
 
-				if errors.Cause(err) != tsdb.ErrNotFound {
-					panic(err)
+		batch := make(commitlog.WritesBatch, len(scrape))
+		for i := 0; i < scrapeCount; i++ {
+			ts += timeDelta
+			id := ident.StringID(string(i))
+			for j, s := range scrape {
+				s.value += 1000
+				batch[j] = commitlog.Write{
+					Series: commitlog.Series{
+						Namespace: defaultM3DBNamespace,
+						ID:        id,
+						TagIter:   s.tags,
+					},
+					Datapoint: m3dbTs.Datapoint{
+						Timestamp: time.Unix(0, ts),
+						Value:     float64(s.value),
+					},
+					Unit:       xtime.Millisecond,
+					Annotation: nil,
 				}
+				// err := b.m3dbStorage.WriteTagged(
+				// 	nil, defaultM3DBNamespace, id, s.tags, time.Unix(0, ts), float64(s.value), xtime.Millisecond, nil)
+				// if err != nil {
+				// 	panic(err)
+				// }
 
-				ref, err := app.Add(s.labels, ts, float64(s.value))
-				if err != nil {
-					panic(err)
-				}
-				s.ref = &ref
+				total++
 			}
 
-			total++
+			err := b.m3dbStorage.WriteTaggedBatch(nil, defaultM3DBNamespace, batch)
```

Comment:
Is this using a development version? The godoc of the package only seems to contain non-batched methods.

Reply:
Yeah sorry I mentioned this in an email to Krasi but should have posted it here. We actually found a bug in our original benchmark that made us much slower and then I quickly hacked together a development branch to improve our efficiency in this type of embedded workload which is what this is running.

I'm currently in the process of productioning those fixes and they should be merged into master within a week. There was basically just some low hanging fruit that we'd never bothered to do because we hadn't had a reason to optimize for the embedded case until now.